import pandas as pd
import matplotlib.pyplot as plt
import sklearn
from Bio.Seq import Seq
from sklearn.svm import OneClassSVM
from scipy.signal import savgol_filter, find_peaks
from sklearn.cluster import DBSCAN
from Bio import pairwise2, motifs
import streamlit as st
import numpy as np
import pdb

def file_path_generation(sample_number):
    """
    This method generates out the file paths for all the files that are part of 1 sample

    :param sample_number: The sample number that needs to be entered to get all Frameworks and all Replicates
    :return: The file paths of all samples
    """

    files = []
    for framework_count in range(1, 4):
        for framework_value in ['A', 'B']:
            files.append(
                "Lymphotrack_Files/" + str(sample_number) + "FR" + str(framework_count) + framework_value + ".csv")
    return files


def create_mother_sequence_mask(mother_sequence):
    """
    Creates a trinucleotide mask for the mother sequence generated by CloneSeer.
    Shifts frame of 6 along the sequence and determines if  position 1 = position 4, position 2 = position 5,
    position 3 = position 6. If it matches, any 0 in the mask at that position turns into a 1

    :param mother_sequence: The most abundant sequence from data
    :return: Mask for mother sequence denoting trinucleotide
    """

    mother_sequence_trinucleotide = np.zeros(len(mother_sequence))
    for character_position in range(len(mother_sequence) - 6):
        if mother_sequence[character_position] == mother_sequence[character_position + 3] \
                and mother_sequence[character_position + 1] == mother_sequence[character_position + 4] \
                and mother_sequence[character_position + 2] == mother_sequence[character_position + 5]:
            for i in range(6):
                mother_sequence_trinucleotide[character_position + i] = 1
    return mother_sequence_trinucleotide


def trinucleotide(data, mother_sequence):
    """
    Creates columns that count the mismatches that occur in trinucleotides, deletions that occur in trinucleotides and
    insertions that occur in trinucleotides.
    :param data: Clone Seer file containing modified sequence, original sequence, number of times observed and inserts
    :param mother_sequence: The mother sequence that is observed most often
    :return: Columns representing mismatches that occur in trinucleotides, deletions that occur in trinucleotides and
    insertions that occur in trinucleotides.
    """
    mother_sequence_trinucleotide = create_mother_sequence_mask(mother_sequence)
    trinucleotide_mismatch_error = []
    dash_mismatch_error = []
    insert_mismatch_error = []
    for index, row in data.iterrows():
        trinucleotide_mismatch_count = 0
        dash_mismatch_count = 0
        insert_count = 0

        # Mismatches and deletions in trinucleotide checked using sequence from Clone Seer. Counts the number of
        # differences that appear between target sequence and mother sequence on a trinucleotide region

        for character_position in range(len(mother_sequence)):
            if row['sequence'][character_position] != mother_sequence[character_position] \
                    and row['sequence'][character_position] != "-" \
                    and mother_sequence_trinucleotide[character_position] == 1:
                trinucleotide_mismatch_count += 1
            if row['sequence'][character_position] != mother_sequence[character_position] and row['sequence'][
                character_position] == "-" and mother_sequence_trinucleotide[character_position] == 1:
                dash_mismatch_count += 1
        trinucleotide_mismatch_error.append(trinucleotide_mismatch_count)
        dash_mismatch_error.append(dash_mismatch_count)

        # Looks through the insert column and count the number of inserts that appear on the trinucleotide position.
        # Inserts at the end of the sequence do not increment insert count due to unreliability of sequencing at the end

        if row['inserts'] == '[]':
            insert_mismatch_error.append(0)
        else:
            insert_positions = row['inserts'][1:-1].split(',')
            for insert_position in insert_positions:
                if int(insert_position) > len(mother_sequence_trinucleotide) - 1:
                    insert_count += 0
                else:
                    if mother_sequence_trinucleotide[int(insert_position)] == 1:
                        insert_count += 1
            insert_mismatch_error.append(insert_count)
    return trinucleotide_mismatch_error, dash_mismatch_error, insert_mismatch_error


def get_total_deletions(data):
    """
    Counts the number of times a deletion occurs in a sequence represented by a dash in the sequence

    :param data: All data points
    :return: A column representing the number of deletions that appear in the sequence generated by Clone seer
    """
    return data.sequence.str.count('-')


def suspicious_mismatch(differences):
    """
    Initially creates a suspicious mismatch to find the max significant homology differences between
    rows immediately above and below. Final Homology difference scans for two consecutive numbers showing up indicating
    possible artifact. Converts point to min difference in order to remove artifact

    :param differences:
    :return: Suspicious mismatch: Greatest distance between the row above and below.
    """
    suspicious_mismatch_number = [abs(differences[0] - differences[1])]
    for differences_position in range(1, len(differences) - 1):
        suspicious_mismatch_number.append(max(abs(differences[differences_position] - differences[differences_position - 1]), abs(differences[differences_position] - differences[differences_position + 1])))
    suspicious_mismatch_number.append(abs(differences[len(differences) - 1] - differences[len(differences) - 2]))

    final_suspicious_mismatch_number = suspicious_mismatch_number.copy()

    for count in range(1, len(suspicious_mismatch_number) - 1):
        if suspicious_mismatch_number[count - 1] == suspicious_mismatch_number[count] and suspicious_mismatch_number[count] > 5:
            if differences[count - 1] > differences[count]:
                final_suspicious_mismatch_number[count] = min(abs(differences[count] - differences[count - 1]), abs(differences[count] - differences[count + 1]))
        if suspicious_mismatch_number[count + 1] == suspicious_mismatch_number[count] and suspicious_mismatch_number[count] > 5:
            if differences[count + 1] > differences[count]:
                final_suspicious_mismatch_number[count] = min(abs(differences[count] - differences[count - 1]), abs(differences[count] - differences[count + 1]))
    return final_suspicious_mismatch_number


def homology_metrics(data, row_of_max_observed, mother_sequence):
    """
    :param data: All data from clone seer
    :param row_of_max_observed: The row with the mother sequence
    :param mother_sequence: The mother sequence representing the sequence that is observed the most
    :return: differences: homology column denoting difference between target sequence and mother sequence
    :return: new_differences: Contiguous blocks of homology differences ex. Diff: [2,3,3,1,0,2,2] Result: [-3, -2, -1, 0, 1]
    """

    # Calculate differences regardless of trinucleotide between target sequence and mother clone
    differences = []
    for index, row in data.iterrows():
        differences.append(sum(a != b for a, b in zip(row['sequence'], mother_sequence)))

    # Calculate differences based upon differences and order. Mother sequence is 0.
    # Getting value from first half
    first_half_value = []
    current_difference = differences[row_of_max_observed]
    count = 0
    for i in range(row_of_max_observed, -1, -1):
        if differences[i] != current_difference:
            count -= 1
            first_half_value.append(count)
            current_difference = differences[i]
        else:
            first_half_value.append(count)
    # Getting Second Half Value
    second_half_value = []
    current_difference = differences[row_of_max_observed]
    count = 0
    for i in range(row_of_max_observed + 1, len(differences)):
        if differences[i] != current_difference:
            count += 1
            second_half_value.append(count)
            current_difference = differences[i]
        else:
            second_half_value.append(count)
    new_difference = first_half_value[::-1] + second_half_value
    suspicious_mismatch_number = suspicious_mismatch(differences)
    return differences, new_difference, suspicious_mismatch_number


def get_index_mismatch(data, mother_sequence):
    """
    Gets the indexes of mismatches between mother clone and the sequence being analyzed
    :param data: Clone seer file
    :param mother_sequence: The most abundant sequence
    :return: A list of list of indexes that are mismatches
    """
    index_mismatch = []
    sequences = data['sequence'].tolist()
    for sequence in sequences:
        index_mismatch.append([i for i in range(len(sequence)) if sequence[i] != mother_sequence[i]])
    return index_mismatch


def is_somatic_hypermutation(data, mother_sequence):
    """
    Gets all of the mismatch indexes between a data sequence and indexes and then shifts frames of 5 in order to find
    the difference between the 1st and 5th indexes to see if it's less than 20

    :param data: All data from clone seer
    :param mother_sequence: The most abundant sequence
    :return: An array of whether or not it is a somatic hypermutation. 0 is not a hypermutation. 1 is a hypermutation
    """

    somatic_hypermutation_result = []
    sequences = data['sequence'].tolist()
    for individual_sequence in sequences:
        mismatch_positions = []
        for sequence_character_position in range(len(individual_sequence)):
            if individual_sequence[sequence_character_position] != mother_sequence[sequence_character_position]:
                mismatch_positions.append(sequence_character_position)
        if len(mismatch_positions) < 5:
            somatic_hypermutation_result.append(0)
        else:
            somatic_hypermutation = 0  # False
            for mismatch_set_of_3_start_index in range(len(mismatch_positions) - 2):
                if mismatch_positions[mismatch_set_of_3_start_index + 2] - mismatch_positions[mismatch_set_of_3_start_index] <= 20:
                    somatic_hypermutation = 1  # True
            somatic_hypermutation_result.append(somatic_hypermutation)
    return somatic_hypermutation_result


def new_metrics(data):
    """
    Generates out more features for the data including deletion count, mismatch, deletion and insertion during trinucleotide,
    homology and contiguous homology differences with mother clone
    :param data: All data from Clone Seer file
    :return: data: Data with new metrics columns attached
    """

    # Mother Clone Identification
    row_of_max_observed = data['numberobserved'].idxmax()
    mother_sequence = data['sequence'][row_of_max_observed]

    # Number of deletions that appear in each sequence
    data['deletion_count'] = get_total_deletions(data)

    data['index_mismatch'] = get_index_mismatch(data, mother_sequence)

    # Mismatches, deletions, and insertions in trinucleotide regions
    data['mismatch_in_trinucleotide'], data['deletion_in_trinucleotide'], data['insertion_in_trinucleotide'] \
        = trinucleotide(data, mother_sequence)
    # Calculate homology differences and contiguous homology block differences
    data['homology_against_mother_clone'], data['contiguous_homology_against_mother_clone'], data[
        'suspicious_mismatch'] = homology_metrics(data, row_of_max_observed, mother_sequence)

    data['somatic_mutations'] = is_somatic_hypermutation(data, mother_sequence)

    return data, row_of_max_observed, mother_sequence


def add_in_outlier_from_suspicious_mismatch(data, outliers_list):
    suspicious_mismatch_results = data['suspicious_mismatch'].tolist()
    
    smoothed_results = savgol_filter(suspicious_mismatch_results, window_length=np.min((len(suspicious_mismatch_results), 5)), polyorder=1)
    peaks, properties = find_peaks(smoothed_results, prominence=1)
    # Generates out plot for analysis
    print(smoothed_results)
    #fig, ax = plt.subplots()
    #ax.pyplot(list(np.arange(0, len(smoothed_results))), smoothed_results)
    #ax.pyplot(list(np.arange(0, len(smoothed_results))), suspicious_mismatch_results)
    #ax.plot(peaks, smoothed_results[peaks], "x")
    #st.write(fig)

    new_outliers_list = outliers_list.copy()

    for individual_peak in peaks:
        search_area = suspicious_mismatch_results[individual_peak - 2: individual_peak + 2]
        max_value = max(search_area)
        max_index = search_area.index(max_value)
        new_outliers_list[(individual_peak + max_index - 2)] = 2
    return new_outliers_list


def outlier_detection(file):
    """
    Takes in a file and finds all the outliers that can be subclones.
    1. Apply DBScan to all sequences except those who have a deletion or insertion on the trinucelotide and find the subclones
    2. Extract all deletions and find average and standard deviation. Deletions who have number observed greater than 1
    standard deviation about average are probable subclones
    3. Extract all insertions and find average and standard deviation. Insertions who have number observed greater than 1
    standard deviation about average are probable subclones
    :param file: File path of clone seer file
    :return: Array filled with likely subclones and mother clone in sequential order
    """

    data = file
    data_with_new_metrics, row_of_mother_clone, mother_clone = new_metrics(data)

    data_without_mother_clone = data_with_new_metrics.drop(index=row_of_mother_clone)
    data_with_insertion = data_without_mother_clone.loc[data_without_mother_clone['insertion_in_trinucleotide'] != 0]
    data_with_deletion = data_without_mother_clone.loc[data_without_mother_clone['deletion_in_trinucleotide'] != 0]
    """
    model_1 = sklearn.cluster.DBSCAN(eps=200, min_samples=10).fit(
        data_without_mother_clone[['numberobserved', 'contiguous_homology_against_mother_clone']])
    #colors_1 = model_1.labels_
    #plt.scatter(data_without_mother_clone["numberobserved"], data_without_mother_clone['insertion_in_trinucleotide'],c=colors_1)
    #plt.xlabel("number of times observed")
    #plt.ylabel("contiguous_homology_against_mother_clone")

    outliers_list = np.zeros(len(data.index))
    outliers_list[row_of_mother_clone] = 1
    outliers_1 = data_without_mother_clone[model_1.labels_ == -1]
    for outlier_count in outliers_1.index.tolist():
        outliers_list[outlier_count] = 1
    """

    outlier_from_number_observed = data_with_new_metrics['numberobserved'].tolist()
    
    smoothed_results = savgol_filter(outlier_from_number_observed, window_length=np.min((len(outlier_from_number_observed),31)), polyorder=1)
    peaks, properties = find_peaks(smoothed_results, prominence=1)
    # Generates out plot for analysis
    # print(smoothed_results)
    # plt.plot(list(np.arange(0, len(smoothed_results))), smoothed_results)
    # plt.plot(list(np.arange(0, len(smoothed_results))), outlier_from_number_observed)
    # plt.plot(peaks, smoothed_results[peaks], "x")
    # plt.show()

    outliers_list = np.zeros(len(data.index))
    outliers_list[row_of_mother_clone] = 1
    for individual_peak in peaks:
        search_area = outlier_from_number_observed[individual_peak - 15: individual_peak + 15]
        max_value = max(search_area)
        max_index = search_area.index(max_value)
        outliers_list[(individual_peak + max_index - 15)] = 1

    df_insert_mean = data_with_insertion['numberobserved'].mean()
    df_insert_std = data_with_insertion['numberobserved'].std()
    df_deletion_mean = data_with_deletion['numberobserved'].mean()
    df_deletion_std = data_with_deletion['numberobserved'].std()
    df_insert_sequenceNotFormatted = data_with_insertion.loc[
        data_with_insertion['numberobserved'] >= df_insert_mean + df_insert_std]
    df_with_deletion_sequenceNotFormatted = data_with_deletion.loc[
        data_with_deletion['numberobserved'] >= df_deletion_mean + 0.5*df_deletion_std]

    for outlier_count in df_insert_sequenceNotFormatted['sequenceNotformatted'].index.tolist():
        outliers_list[outlier_count] = 1
    for outlier_count in df_with_deletion_sequenceNotFormatted['sequenceNotformatted'].index.tolist():
        outliers_list[outlier_count] = 1
    
    outliers_list = add_in_outlier_from_suspicious_mismatch(data_with_new_metrics, outliers_list)
    data['outliers'] = outliers_list
    outlier_sequential = []
    outlier_type = []
    for position in range(len(outliers_list)):
        if outliers_list[position] == 1:
            outlier_sequential.append(data['sequenceNotformatted'].iloc[position])
            outlier_type.append(1)
        if outliers_list[position] == 2:
            outlier_sequential.append(data['sequenceNotformatted'].iloc[position])
            outlier_type.append(2)
    return data, outlier_sequential, outlier_type


def partition(data, outliers, outlier_type):
    """
        Takes the data and for each sequence assigns a set of alignment score (one against each outlier) and then picks the one
        that has the least difference between the sequence length and the target sequence.
        :param data: Clone Seer Data
        :param outliers: All of the outliers
        :param outlier_type: The outlier type. 1 = Outlier from trinucleotide or num observed. 2 = somatic hypermutation
        :return: Data with a column for closest match
    """
    sequences = data['sequenceNotformatted'].tolist()

    # First Pass utilizing only outliers coming from num observed filter
    initial_outlier_class = [i for i, x in enumerate(outlier_type) if x == 1]  # Gets all of the indexes of outliers that came from the original num observed filter
    initial_outlier_class_corresponding_sequence = [] # Corresponding sequence to num observed filter outliers
    for individual_outlier_class in initial_outlier_class:
        initial_outlier_class_corresponding_sequence.append(outliers[individual_outlier_class])
    initial_outlier_index = []  # Corresponding Outlier Index Value in the actual outliers
    for individual_outlier in initial_outlier_class_corresponding_sequence:
        initial_outlier_index.append(data.index[data['sequenceNotformatted'] == individual_outlier])

    # First pass matching
    closest_match = []
    count = 0
    for sequence in sequences:
        outlier_score = len(sequence)
        best_outlier_position = -1
        outlier_seq_pos = 2 * len(sequences)
        sequence_no_dash = sequence.replace("-", "")
        
        for outlier_count in range(len(initial_outlier_class_corresponding_sequence)):
            alignments = pairwise2.align.globalms(sequence_no_dash, initial_outlier_class_corresponding_sequence[outlier_count].replace("-", ""), 1, -1, -1, 0, score_only=True)
            if len(sequence_no_dash) - alignments < outlier_score:
                outlier_score = len(sequence_no_dash) - alignments
                best_outlier_position = initial_outlier_class[outlier_count]
                outlier_seq_pos = initial_outlier_index[outlier_count]
            if len(sequence_no_dash) - alignments == outlier_score:
                if outlier_seq_pos - count > initial_outlier_class[outlier_count] - count:
                    best_outlier_position = initial_outlier_class[outlier_count]
                    outlier_seq_pos = initial_outlier_class[outlier_count]
        closest_match.append(best_outlier_position)
        count += 1

    # Grabbing original index for second pass. Takes 20 above and 20 below for each outlier and sees if class needs to be adjusted with better score
    outlier_index = []  # Corresponding Outlier Index Value in the actual outliers
    for individual_outlier in outliers:
        outlier_index.append(data.loc[data['sequenceNotformatted'] == individual_outlier].index.values)
    outlier_class = 0
    for individual_outlier_index in outlier_index:
        for search_area in range(max(0, individual_outlier_index[0] - 20), min(individual_outlier_index[0] + 21, len(sequences))):
            new_score = pairwise2.align.globalms(sequences[search_area].replace("-", ""), sequences[individual_outlier_index[0]].replace("-", ""), 1, -1, -1, 0, score_only=True)
            old_score = pairwise2.align.globalms(sequences[search_area].replace("-", ""), outliers[closest_match[search_area]].replace("-", ""), 1, -1, -1, 0, score_only=True)
            if new_score > old_score:
                closest_match[search_area] = outlier_class
        outlier_class += 1
    data['closest match'] = closest_match
    return data


def partitionV2(data, outliers, outlier_type, variantoutliers):
    """
        Takes the data and for each sequence assigns a set of alignment score (one against each outlier) and then picks the one
        that has the least difference between the sequence length and the target sequence.
        :param data: Clone Seer Data
        :param outliers: All of the outliers
        :param outlier_type: The outlier type. 1 = Outlier from trinucleotide or num observed. 2 = somatic hypermutation
        :param variantoutliers: All the outliers found by trinucleotide but trimmed to variants
        :return: Data with a column for closest match
    """
    sequences_pass1 = data['variantseq'].tolist()
    sequences = data['sequenceNotformatted'].tolist()
    # First Pass utilizing only outliers coming from num observed filter
    initial_outlier_class = [i for i, x in enumerate(outlier_type) if x == 1]  # Gets all of the indexes of outliers that came from the original num observed filter
    initial_outlier_class_corresponding_sequence = []  # Corresponding sequence to num observed filter outliers
    for individual_outlier_class in initial_outlier_class:
        initial_outlier_class_corresponding_sequence.append(outliers[individual_outlier_class])
    initial_outlier_index = []  # Corresponding Outlier Index Value in the actual outliers
    for individual_outlier in initial_outlier_class_corresponding_sequence:
        initial_outlier_index.append(data.index[data['sequenceNotformatted'] == individual_outlier])

    # First pass matching
    closest_match = []
    count = 0
    for sequence in sequences_pass1:
        outlier_score = len(sequence)
        best_outlier_position = -1
        outlier_seq_pos = 2 * len(sequences_pass1)
        sequence_no_dash = sequence.replace("-", "-")
        for outlier_count in range(len(variantoutliers)):
            alignments = pairwise2.align.globalms(sequence_no_dash,
                                                  variantoutliers[outlier_count].replace(
                                                      "-", "-"), 1, -1, -1, 0, score_only=True)

            if alignments != []:                                        
                if len(sequence_no_dash) - alignments < outlier_score:
                    outlier_score = len(sequence_no_dash) - alignments
                    best_outlier_position = initial_outlier_class[outlier_count]
                    outlier_seq_pos = initial_outlier_index[outlier_count]
                if len(sequence_no_dash) - alignments == outlier_score:
                    if outlier_seq_pos - count > initial_outlier_class[outlier_count] - count:
                        best_outlier_position = initial_outlier_class[outlier_count]
                        outlier_seq_pos = initial_outlier_class[outlier_count]
        closest_match.append(best_outlier_position)
        count += 1

    # Grabbing original index for second pass. Takes 20 above and 20 below for each outlier and sees if class needs to be adjusted with better score
    outlier_index = []  # Corresponding Outlier Index Value in the actual outliers
    for individual_outlier in outliers:
        outlier_index.append(data.loc[data['sequenceNotformatted'] == individual_outlier].index.values)
    outlier_class = 0
    for individual_outlier_index in outlier_index:
        for search_area in range(max(0, individual_outlier_index[0] - 20),
                                 min(individual_outlier_index[0] + 21, len(sequences))):
            new_score = pairwise2.align.globalms(sequences[search_area].replace("-", ""),
                                                 sequences[individual_outlier_index[0]].replace("-", ""), 1, -1, -1, 0,
                                                 score_only=True)
            old_score = pairwise2.align.globalms(sequences[search_area].replace("-", ""),
                                                 outliers[closest_match[search_area]].replace("-", ""), 1, -1, -1, 0,
                                                 score_only=True)
            if new_score > old_score:
                closest_match[search_area] = outlier_class
        outlier_class += 1
    data['closest match'] = closest_match
    return data


def process_fasta_from_ebi(fasta_file_path, data):
    """
    The function takes the fasta_file generated from EBI listing the V genes and matches it up against the outliers
    :param fasta_file_path: The path to the fasta file containing all of the genes
    :param data: Data with all of the metrics
    :return: Data that has a new column denoting V-gene
    """

    # Builds the IGHV gene database from the EBI file
    fasta_file = open(fasta_file_path, "r")
    ighv_database = {}
    current_key = ""
    current_sequence = ""
    for line in fasta_file:
        if line[0] == '>':
            split_title_line = line.split("|")
            ighv_database[current_key] = current_sequence.replace("\n", "").replace(".", "").upper()
            current_key = split_title_line[1]
            current_sequence = ""
        else:
            current_sequence = current_sequence + line
    del ighv_database[""] # Needs to delete first item in order to place

    corresponding_ighv = []
    corresponding_score = []
    data_sequences = data['sequence']
    data_outlier = data['outliers']

    for individual_sequence_position in range(len(data_sequences)):
        if data_outlier[individual_sequence_position] == 1 or data_outlier[individual_sequence_position] == 2:
            max_score = 0
            max_score_key = ""
            for key in ighv_database.keys():
                alignment_score = pairwise2.align.localms(data_sequences[individual_sequence_position], ighv_database[key], 1, -1, -1, 0, score_only=True)
                if float(alignment_score) / float(len(ighv_database[key])) > max_score:
                    max_score = float(alignment_score) / float(len(ighv_database[key]))
                    max_score_key = key
            corresponding_ighv.append(max_score_key)
            corresponding_score.append(max_score)
        else:
            corresponding_ighv.append('Not Outlier')
            corresponding_score.append(0)
    data['v-genes'] = corresponding_ighv
    data['v-genes score'] = corresponding_score
    return data


def KevinsFunction(table_of_sequences):
    table_of_sequences['inserts']=table_of_sequences['inserts'].apply(lambda x: str(x))
    data, found_outlier, outlier_type = outlier_detection(table_of_sequences)
    outlierseq = data[data[['outliers', 'somatic_mutations']].apply(lambda x: x[0] == 1 and x[1] == 0, axis=1)].sequence.apply(lambda x: Seq(x))
    outlierseqlist = outlierseq.tolist()
    print(outlierseqlist)
    
    outlierseqmotif = motifs.create(outlierseqlist,alphabet='ACGT-')
    motifcount= pd.DataFrame(outlierseqmotif.counts).apply(lambda x: sum(x), axis=1)[0]
    
    mutated_positions = np.where([not (a == motifcount or c == motifcount or t == motifcount or g == motifcount) for a, c, g, t in zip(outlierseqmotif.counts['A'], outlierseqmotif.counts['C'], outlierseqmotif.counts['G'], outlierseqmotif.counts['T'])])[0]
    seq_onlymutated = data.sequence.apply(lambda x: ''.join([x[i] for i in mutated_positions]))
    
    
    #if len(outlierseq)>1:
    #    outlierseqmutated = outlierseq.apply(lambda x: ''.join([str(x)[i] for i in mutated_positions])).tolist()
    #    outlierseqmutated = [i for i in outlierseqmutated if i.count('-')!=len(i)]
    #    data['variantseq'] = seq_onlymutated
    #else:
    #    outlierseqmutated = outlierseqlist
    #    data['variantseq'] =data.sequence

    outlierseqmutated = outlierseq.apply(lambda x: ''.join([str(x)[i] for i in mutated_positions])).tolist()
    data['variantseq'] = seq_onlymutated
    v_gene_data = process_fasta_from_ebi('Vs.fasta', data)
    
    partitioned_data = partitionV2(v_gene_data, found_outlier, outlier_type, outlierseqmutated)
    
    return(partitioned_data)